### PR TITLE
meta: make auto increment id can be adjust.

### PR DESCRIPTION
**Pull request by imtbkcat**
### What problem does this PR solve? 
Make auto increment id step can be adjust automatically according to auto id request speed.

### What is changed and how it works?
add function `NextStep` which can produce the next auto id step according to previous step.

There are still many test need to be change, so do not merge.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

 - Possible performance regression
 - Increased code complexity

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: meta/autoid/autoid.go [hunk 1]
```diff
@@ -214,14 +223,17 @@ func (alloc *allocator) alloc4Unsigned(tableID int64) (int64, error) {
 	if alloc.base == alloc.end { // step
 		var newBase, newEnd int64
 		startTime := time.Now()
+		consumeDur := startTime.Sub(alloc.lastAllocTime)
+		alloc.step = NextStep(alloc.step, consumeDur)
+		alloc.lastAllocTime = startTime
```
Comment (coocood) [inline meta/autoid/autoid.go]: We should store the time after we run the transaction.

**Comment by coocood** [comment]
@imtbkcat 
We need unit test for `NextStep`.

File: meta/autoid/autoid.go [hunk 2]
```diff
@@ -29,6 +30,12 @@ import (
 	"go.uber.org/zap"
 )
 
+const (
+	minStep            = 5000
```
Comment (coocood) [inline meta/autoid/autoid.go]: Can we make minStep lower like 1000?

Comment (imtbkcat) [reply meta/autoid/autoid.go]: yes

File: meta/autoid/autoid.go [hunk 3]
```diff
@@ -289,12 +304,35 @@ func (alloc *allocator) Alloc(tableID int64) (int64, error) {
...
 
+// NextStep return new auto id step according to previous step and consuming time.
+func NextStep(curStep int64, consumeDur time.Duration) int64 {
+	failpoint.Inject("mockAutoIDChange", func(val failpoint.Value) {
+		if val.(bool) {
+			failpoint.Return(step)
+		}
+	})
+
+	if consumeDur.Nanoseconds() == 0 {
```
Comment (coocood) [inline meta/autoid/autoid.go]: When will consumeDur be 0?

File: meta/autoid/autoid.go (hunk 2 above)
Comment (ngaut) [reply meta/autoid/autoid.go]: 1000 is too small, let's say we want 200000 insert QPS in a single tidb server, how many transaction do we need? what if we met transaction conflict while updating the allocated meta.

File: meta/autoid/autoid.go [hunk 4]
```diff
@@ -289,12 +304,35 @@ func (alloc *allocator) Alloc(tableID int64) (int64, error) {
...
+		return curStep
+	}
+	y := defaultComsumeTime.Seconds() / consumeDur.Seconds()
```
Comment (coocood) [inline meta/autoid/autoid.go]: We need a better name than `y`.

File: meta/autoid/autoid.go (hunk 2 above)
Comment (coocood) [reply meta/autoid/autoid.go]: The `minStep` only matters when the insert QPS is very low.

File: ddl/db_test.go [hunk 5]
```diff
@@ -2170,6 +2171,10 @@ func (s *testDBSuite4) TestComment(c *C) {
 }
 
 func (s *testDBSuite5) TestRebaseAutoID(c *C) {
```
Comment (jackysp) [inline ddl/db_test.go]: It is better to move them to the same suite.

**Comment by coocood** [comment]
LGTM

File: meta/autoid/autoid.go [hunk 6]
```diff
@@ -29,6 +30,12 @@ import (
 	"go.uber.org/zap"
 )
 
+const (
+	minStep            = 1000
+	maxStep            = 2000000
+	defaultComsumeTime = 10 * time.Second
```
Comment (lamxTyler) [inline meta/autoid/autoid.go]: `Consume`?

**Pull request merged**